### PR TITLE
executor, util: fix UnionScan Next() skip reading data when passed chunk capacity is 0

### LINKED ISSUE #36903: TiDB (with tikv) produces different results for SELECT statements when using transaction
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

**Setup the environment:**

```
tiup playground &
mysql -h "127.0.0.1" -u root -P 4000
MySQL> create database testdb;
MySQL> \q
mysql -h "127.0.0.1" -u root -P 4000 -D testdb < mysql_bk.sql
```

mysql_bk.sql: [mysql_bk.txt](https://github.com/pingcap/tidb/files/9259646/mysql_bk.txt)

**Testcase 1**
```
mysql -h "127.0.0.1" -u root -P 4000 -D testdb

mysql> insert into t_vwvgdc (wkey, pkey, c_rdsfbc) values (155, 228000, 99.50);
mysql> select
                  ref_4.pkey as c0
                from
                  t_vwvgdc as ref_4
                where 0 <> 0
              union
              select
                  ref_5.pkey as c0
                from
                  t_vwvgdc as ref_5;
```

**Testcase 2**
```
mysql -h "127.0.0.1" -u root -P 4000 -D testdb

mysql> BEGIN OPTIMISTIC;
mysql> insert into t_vwvgdc (wkey, pkey, c_rdsfbc) values (155, 228000, 99.50);
mysql> select
                  ref_4.pkey as c0
                from
                  t_vwvgdc as ref_4
                where 0 <> 0
              union
              select
                  ref_5.pkey as c0
                from
                  t_vwvgdc as ref_5;
mysql> commit;
```

### 2. What did you expect to see? (Required)

The execution result of SELECT statements in Testcase 1 and Testcase 2 are same. 

### 3. What did you see instead (Required)

**Output of SELECT statement in Testcase 1**
```
+--------+
| c0     |
+--------+
|  17000 |
|  15000 |
|  66000 |
|    ... |
| 228000 |
+--------+
30 rows in set (0.00 sec)
```

**Output of SELECT statement in Testcase 2**
```
Empty set (0.00 sec)
```

Their results are different.

### 4. What is your TiDB version? (Required)

```
Release Version: v6.1.0
Edition: Community
Git Commit Hash: 1a89decdb192cbdce6a7b0020d71128bc964d30f
Git Branch: HEAD
UTC Build Time: 2022-08-01 09:18:07
GoVersion: go1.18
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #36903

Problem Summary:

### What is changed and how it works?

**In the old UnionScan executor code, if an empty chk  is passed to `UnionScan.Next(chk)` , the UnionScan executor will not read any data.**

This bug could be trigger in `select pkey from t_vwvgdc where 0 <> 0 union select pkey from t_vwvgdc;`
The union first children return empty chunk, and the chunk is reused for the second children (union scan) to read data.

```
mysql> explain select pkey from t_vwvgdc where 0 <> 0 union select pkey from t_vwvgdc;
+------------------------------+---------+-----------+----------------+-------------------------------------------------------+
| id                           | estRows | task      | access object  | operator info                                         |
+------------------------------+---------+-----------+----------------+-------------------------------------------------------+
| HashAgg_13                   | 24.20   | root      |                | group by:Column#7, funcs:firstrow(Column#7)->Column#7 |
| └─Union_14                   | 29.00   | root      |                |                                                       |
|   ├─TableDual_16             | 0.00    | root      |                | rows:0                                                |
|   └─UnionScan_18             | 29.00   | root      |                |                                                       |
|     └─TableReader_20         | 29.00   | root      |                | data:TableFullScan_19                                 |
|       └─TableFullScan_19     | 29.00   | cop[tikv] | table:t_vwvgdc | keep order:false, stats:pseudo                        |
+------------------------------+---------+-----------+----------------+-------------------------------------------------------+
6 rows in set (0.00 sec)
```

Now I change `req.GrowAndReset(us.maxChunkSize)` to ensure capacity > 0 for the `req`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix a bug that UnionScan's Next() function skips reading data when the passed chunk's capacity is 0
```

### CODE REVIEW

File: util/chunk/chunk.go [hunk 1]
```diff
@@ -320,6 +320,9 @@ func reCalcCapacity(c *Chunk, maxChunkSize int) int {
 	if c.NumRows() < c.capacity {
 		return c.capacity
 	}
+	if c.capacity == 0 {
+		return InitialCapacity
```
Comment (xhebox) [inline util/chunk/chunk.go:324]: IMO, this is better that it does not bypass the max chunk size check:

```
if c.capacity == 0 {
		c.capacity = InitialCapacity
} else {
        c.capacity *= 2
}
return mathutil.Min(c.capacity, maxChunkSize)
```

Comment (tiancaiamao) [reply util/chunk/chunk.go:324]: Done

Comment (tiancaiamao) [reply util/chunk/chunk.go:326]: Well, I find this function `reCalcCapacity` is side-effect free, i.e.  it just calculate the new capacity, without change `c.capacity`

Comment (tiancaiamao) [reply util/chunk/chunk.go:326]: @xhebox

Comment (xhebox) [reply util/chunk/chunk.go:326]: OK, what about a new variable, or just keep it as-is. I am fine with both.

**Review by xhebox — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**